flink-connector-aws/flink-connector-aws-kinesis-streams/src/main/java/org/apache/flink/connector/kinesis/source/metrics/KinesisSourceShardMetrics.java

```diff
@@ -0,0 +1,104 @@
+import java.util.Map;
+
+/**
+ * A utility class for handling metrics in {@link
+ * org.apache.flink.connector.kinesis.source.reader.KinesisStreamsSourceReader}.
+ */
+@Internal
+public class KinesisSourceShardMetrics {
+    private static final Logger LOG = LoggerFactory.getLogger(KinesisSourceShardMetrics.class);
+
+    private final Map<String, Long> millisBehindLatestMap = new HashMap<>();
+    private final Map<String, MetricGroup> shardMetricGroupMap = new HashMap<>();
+    private final MetricGroup sourceReaderMetricGroup;
+
+    public KinesisSourceShardMetrics(final MetricGroup metricGroup) {
+        this.sourceReaderMetricGroup = metricGroup;
+    }
+
+    public void registerShardMetricGroup(KinesisShardSplit shard) {
+        LOG.debug("Registering the metric group for shard {}", shard.getShardId());
+        Arn shardArn = Arn.fromString(shard.getStreamArn());
+
+        millisBehindLatestMap.put(shard.getShardId(), -1L);
+        final MetricGroup shardGroup =
+                this.sourceReaderMetricGroup
+                        .addGroup(
+                                MetricConstants.ACCOUNT_ID_METRIC_GROUP, shardArn.accountId().get())
+                        .addGroup(MetricConstants.REGION_METRIC_GROUP, shardArn.region().get())
+                        .addGroup(
+                                MetricConstants.STREAM_METRIC_GROUP, shardArn.resource().resource())
+                        .addGroup(MetricConstants.SHARD_METRIC_GROUP, shard.getShardId());
+
+        shardGroup.gauge(
+                MetricConstants.MILLIS_BEHIND_LATEST_GAUGE,
+                () -> millisBehindLatestMap.get(shard.getShardId()));
+        shardMetricGroupMap.put(shard.getShardId(), shardGroup);
+
+        LOG.debug(
+                "Registered metric identifier {}",
+                shardGroup.getMetricIdentifier(MetricConstants.MILLIS_BEHIND_LATEST_GAUGE));
+    }
+
+    public void updateShardMillisBehindLatest(
+            KinesisShardSplitState shardState, Long millisBehindLatest) {
+        LOG.debug(
+                "Updating millisBehindLatest metric for shard {} to {}",
+                shardState.getShardId(),
+                millisBehindLatest);
+        millisBehindLatestMap.put(shardState.getShardId(), millisBehindLatest);
+    }
+
+    public void unregisterShardMetricGroup(KinesisShardSplitState shardState) {
+        if (shardMetricGroupMap.containsKey(shardState.getShardId())) {
+            MetricGroup shardMetricGroup = shardMetricGroupMap.get(shardState.getShardId());
+
+            if (shardMetricGroup instanceof AbstractMetricGroup) {
+                ((AbstractMetricGroup<?>) shardMetricGroup).close();
+            }
+
+            shardMetricGroupMap.remove(shardState.getShardId());
+        } else {
+            LOG.warn(
+                    "Shard metric group unregister failed. Metric group for {} does not exist.",
+                    shardState.getShardId());
+        }
+    }
+
+    @VisibleForTesting
```

Comment:
This is not in line with Flink's testing principles. Is there a reason we cannot set this up using a `TestMetricGroup` instead?